### DISCUSSION TITLE
How to write an asset that depends on an earlier partition of itself

### DISCUSSION THREAD

**Discussion by sryza** [dagster-io/dagster#11829] [Q&A]
You can make an asset depend on earlier partitions of itself by using a `TimeWindowPartitionMapping` with a `start_offset` and `end_offset`.

```python
from dagster import asset, DailyPartitionsDefinition, TimeWindowPartitionMapping, AssetDep


daily_partitions_def = DailyPartitionsDefinition(start_date="2020-01-01")
depend_on_past = TimeWindowPartitionMapping(start_offset=-1, end_offset=-1)


@asset(
    partitions_def=daily_partitions_def,
    deps=[AssetDep("asset1", partition_mapping=depend_on_past)],
)
def asset1() -> None:
    ...
```

**Comment by yeachan153** [comment]
Thanks @sryza, I also like that is says that "this asset depends on earlier partitions of itself" in the definitions page. 

I'm trying this out and running into a couple of issues. We mainly use graph backed assets, so I tried this:

```python
@graph(out={"etl_1_test_graph_asset": GraphOut()})
def etl_1_test_graph_asset(r_etl_1_test_graph_asset):
    r_test_sleep_1 = test_sleep_1(r_etl_1_test_graph_asset)
    r_end = end(r_test_sleep_1)

    return {"etl_1_test_graph_asset": r_end}

r_etl_1_test_graph_asset = AssetsDefinition.from_graph(
    etl_1_test_graph_asset,
    keys_by_input_name={"r_etl_1_test_graph_asset": AssetKey("etl_1_test_graph_asset")},
    partition_mappings={"r_etl_1_test_graph_asset": TimeWindowPartitionMapping(start_offset=-1, end_offset=-1)},
    partitions_def=HourlyPartitionsDefinition("2023-01-20-10:00")
)

r_etl_1_test_graph_asset_job = define_asset_job(
    name="etl_1_test_graph_asset_job",
    selection=AssetSelection.assets(r_etl_1_test_graph_asset),
    partitions_def=HourlyPartitionsDefinition("2023-01-20-10:00"), # hourly
)

etl_1_test_graph_asset_job_schedule = build_schedule_from_partitioned_job(r_etl_1_test_graph_asset_job)
```

Somehow the scheduler still seems to start a new run every partition even though the last partition never materialised successfully. Am I doing something wrong? I also noticed the same behaviour with the example you shared:

<img width="920" alt="image" src="https://user-images.githubusercontent.com/30076090/213937693-b957dd4b-a64c-4ac5-a3a3-4744da8f1c45.png">

**Reply by sryza** [reply]
Currently, you need to use `build_asset_reconciliation_sensor` if you want it to automatically respect the self-partition dependency when submitting runs.

**Reply by yeachan153** [reply]
Ah OK. Is there a plan to add support for this for schedules, or do you not see a reason to do so since Dagster also seems to treat sensors as first class components responsible for scheduling behaviour?

I've added the sensor and turned the schedule off:
```python
r_etl_1_test_graph_asset_job_sensor = build_asset_reconciliation_sensor(
    asset_selection=AssetSelection.assets(r_etl_1_test_graph_asset),
    name="etl_1_test_graph_asset_job_sensor",
    default_status=DefaultSensorStatus.RUNNING,
)
```

I can't seem to get any runs to start though. I initially assumed the first one partition would just start since there would be no materialisations in the upstream partition. Since that didn't happen, I tried backfilling a partition to see if the next one would get kicked off, but that also didn't work for some reason.

[edit]: I'm quite confused now. I pulled the start date of my asset forward by a couple of days to today. The asset now has a first partition of `2023-01-22-18:00`:
<img width="344" alt="image" src="https://user-images.githubusercontent.com/30076090/213941903-a702aad4-a800-411e-b796-8abdb8ea1f7e.png">

But the sensor is launching runs for the whole day in batches (runs failing on purpose):
<img width="1221" alt="image" src="https://user-images.githubusercontent.com/30076090/213941993-be8b74bf-9a0c-43cf-ae84-53c77df2d95b.png">

**Reply by sryza** [reply]
> Ah OK. Is there a plan to add support for this for schedules, or do you not see a reason to do so since Dagster also seems to treat sensors as first class components responsible for scheduling behaviour?

I think it's worth investigating adding this for schedules.  I filed https://github.com/dagster-io/dagster/issues/11833 to track this.

> But the sensor is launching runs for the whole day in batches (runs failing on purpose):

Hmm, that's unexpected that multiple would get requested at once if you have a self-partition dep. Do you have a code snippet I'd be able to try this out with? I just tried with this and got the expected behavior (one run per sensor tick):
```python
from dagster import (
    asset,
    AssetIn,
    HourlyPartitionsDefinition,
    TimeWindowPartitionMapping,
    Nothing,
    build_asset_reconciliation_sensor,
    AssetSelection,
    Definitions,
)


@asset(
    partitions_def=HourlyPartitionsDefinition(start_date="2023-01-22-00:00"),
    ins={
        "asset1": AssetIn(
            partition_mapping=TimeWindowPartitionMapping(start_offset=-1, end_offset=-1),
            dagster_type=Nothing,
        )
    },
)
def asset1() -> None:
    ...


defs = Definitions(
    assets=[asset1], sensors=[build_asset_reconciliation_sensor(AssetSelection.all())]
)
```

It's possible that we have a bug related graph-backed assets?

Btw, I think the reason that you weren't seeing any runs until you pulled the start date up is that we won't fill in partitions until all their ancestors are filled. Is that undesirable behavior?  We also don't kick off runs for partitions more than 24 hours earlier than the current date, to avoid unintentionally backfilling enormous numbers of partitions.

**Comment by yeachan153** [comment]
Thanks @sryza, sure, here's a code snippet. Seems like depends on previous partition is not being respected, and identical partitions are also being run again.

```python
import random
import time
from dagster import (
    In, Nothing, graph, op, GraphOut,
    AssetsDefinition, AssetKey, TimeWindowPartitionMapping,
    HourlyPartitionsDefinition, build_asset_reconciliation_sensor,
    AssetSelection, DefaultSensorStatus, repository
)

@op(ins={"start": In(Nothing)})
def test_sleep_1():
    if random.random() > 0.2:
        raise RuntimeError()
    time.sleep(1)

@op(ins={"start": In(Nothing)})
def end():
    time.sleep(1)

@graph(out={"etl_1_test_graph_asset": GraphOut()})
def etl_1_test_graph_asset(r_etl_1_test_graph_asset):
    r_test_sleep_1 = test_sleep_1(r_etl_1_test_graph_asset)
    r_end = end(r_test_sleep_1)
    return {"etl_1_test_graph_asset": r_end}

r_etl_1_test_graph_asset = AssetsDefinition.from_graph(
    etl_1_test_graph_asset,
    keys_by_input_name={"r_etl_1_test_graph_asset": AssetKey("etl_1_test_graph_asset")},
    partition_mappings={"r_etl_1_test_graph_asset": TimeWindowPartitionMapping(start_offset=-1, end_offset=-1)},
    partitions_def=HourlyPartitionsDefinition("2023-01-22-18:00")
)

r_etl_1_test_graph_asset_job_sensor = build_asset_reconciliation_sensor(
    asset_selection=AssetSelection.assets(r_etl_1_test_graph_asset),
    name="etl_1_test_graph_asset_job_sensor",
    default_status=DefaultSensorStatus.RUNNING,
)

@repository
def test_repo():
    return [r_etl_1_test_graph_asset, r_etl_1_test_graph_asset_job_sensor]
```

> Btw, I think the reason that you weren't seeing any runs until you pulled the start date up is that we won't fill in partitions until all their ancestors are filled. Is that undesirable behavior? We also don't kick off runs for partitions more than 24 hours earlier than the current date, to avoid unintentionally backfilling enormous numbers of partitions.

OK, thanks for clarifying! So does that mean for the sensor to start working where the asset has a start date of over 24 hours ago, I'd have to backfill until 24 hours before the current time in order for the sensor to start kicking off runs? If so, that might be a bit unexpected. Off the top of my head, it would probably surprise me least if the sensor behaved similarly to the schedule in the sense that it just applied to partitions going forwards, but not backwards regardless of the start time of the asset.

One thing I noticed was that the sensor was also creating runs for the whole 24 hours, although the asset partitions started at 18:00. That may be a bug?

**Reply by sryza** [reply]
> One thing I noticed was that the sensor was also creating runs for the whole 24 hours, although the asset partitions started at 18:00. That may be a bug?

Ah, I think this is the problem.  @OwenKephart it looks like this might be a bug introduced in [this change](https://github.com/dagster-io/dagster/pull/11755/files)? In `get_never_requested_never_materialized_partitions`, if the allowable time window starts before the start date of the PartitionsDefinition, it looks like we end up adding requesting partitions before the start date.

**Reply by yeachan153** [reply]
OK, happy to go ahead and make an issue if you want?

Regarding identical partitions being run multiple times, that seems like a separate issue. Were you able to replicate that with the snippet I shared? That sounds like it could be a different issue.

**Reply by sryza** [reply]
> Regarding identical partitions being run multiple times, that seems like a separate issue. Were you able to replicate that with the snippet I shared? That sounds like it could be a different issue.

Although it's a different symptom, I suspect it's the same root issue in this case - I think we're not successfully marking those partitions before the start date as requested because they don't "exist" in the first place. So then they get requested again.

**Reply by sryza** [reply]
@OwenKephart is working on a fix here: https://github.com/dagster-io/dagster/pull/11842/files

**Reply by yeachan153** [reply]
Thanks both, that's now working as expected!

**Reply by yeachan153** [reply]
One thing I'm unsure about is whether the sensor should be backfilling at all. It feels a bit arbitrary that 24 hours was chosen as the time frame which determines whether the sensor starts backfilling or not triggering runs at all. I see two potential issues:

- To get the sensor to work, we'd need to backfill up until 24 hours before, but remember not to include the 23 partitions within 24 hours if they want to avoid launching duplicate runs from both the sensor and the backfill in case we have hourly partitions. This might be easy to get caught out on.
- This behaviour is different for daily partitions, where we'd just have to not include the last day's partition in the backfill.

Maybe there were other reasons 24 hours was chosen?

It might be simpler if the behaviour was changed so that:

- For both hourly/daily partitions, if the start date/hour is the last partition, the sensor starts a run. E.g. an asset with a start date of `2023-01-25-00:00` would immediately start running if I deployed the sensor at `2023-01-25-01:10`. Similarly, a daily asset with a start date of `2023-01-24` would immediately start running if I deployed the sensor on `2023-01-25`.
- If the start date falls outside, e.g. an hourly asset has a start date of `2023-01-25-00:00` and I deployed the sensor on `2023-01-25-02:00`, the sensor does not create runs until all currently applicable partitions have been backfilled. Same for the daily partition.

This might be better as an issue, I can create one and we can discuss there if you'd like?

**Reply by sryza** [reply]
@OwenKephart - thoughts on @yeachan153's first question above? We could just have the reconciliation sensor backfill the most recent partition? I feel like we discussed that but there was some reason we thought it might cause difficulties?  Another thing we could do is say that the reconciliation sensor shouldn't backfill any asset partitions that are targeted by an active backfill.  I suspect we'll need to do that for a number of other situations as well.

To your second question @yeachan153, I think that would be fairly difficult with our current architecture. I'm having trouble thinking of a great workaround.

**Reply by OwenKephart** [reply]
Hi @yeachan153 ! I appreciate your detailed feedback here -- to be honest, there isn't anything particularly special about the 1 day limit (we were mostly just trying to set some sort of boundary to avoid launching huge numbers of runs), and just allowing it to fill in the most recent partition does make a good amount of sense.

My main concern with that approach is around users that have high-frequency partitions (e.g. 15 minutes). If you have a long chain of assets, each partitioned by 15-minute windows, then it's possible that it could take more than 15 minutes to fill in a partition at the end of that chain, making it impossible for the reconciliation sensor to "do its job" for those assets later in the graph.

My reasoning was that a 1-day limit could gloss over these types of issues (it's rare for a computation to require more than 1 day to percolate through the graph), but in the end it was an arbitrary choice. In reality, we do want to provide users the ability to be more specific about their desires here (i.e. some sort of ReconciliationCondition object that you can provide to the asset definition to inform the sensor when it is "allowed" to materialize a given asset/partition).

I'm fairly convinced by your argument that the default behavior (in a world where you can change the default) should be "just fill in the most recent partition", but I'm not confident about making that change before things are configurable.

**Reply by yeachan153** [reply]
Hey @OwenKephart, thanks for your input.

My personal feeling is that it is the user's responsibility to ensure any chain of assets running on a particular partition window are able to finish within that window. I don't think it is really possible for Dagster to take on that responsibility, and the 1-day limit is now a will work for some/won't work for some solution which also results in unexpected behaviour elsewhere. That doesn't seem ideal.

What would be good to have is a set of robust monitoring tools that helps with operational work. Some examples:
- Explainable sensors: why is this sensor not generating a run request? Why did it generate a run request? Right now, it's hard to understand. E.g. if there were 5 dependencies and only one is still waiting, all the user sees is a "Skipped" message. Maybe in addition, to what assets does this sensor apply to?
- Back to the point above, an easy way to check which assets are falling behind their partition window. Users can then fix this so that their code doesn't take so long.

**Comment by renzhe-brian** [comment]
Hi @sryza , @OwenKephart . This is a follow up for Yeachan's question. We are currently using dagser version 1.3.9. I copied @yeachan153 's code snippet and host dagster locally to test the self dependant behavior:
```python
import random
import time
from dagster import (
    In, Nothing, graph, op, GraphOut,
    AssetsDefinition, AssetKey, TimeWindowPartitionMapping,
    HourlyPartitionsDefinition, build_asset_reconciliation_sensor,
    AssetSelection, DefaultSensorStatus, repository
)

@op(ins={"start": In(Nothing)})
def test_sleep_1():
    # if random.random() > 0.2:
    #     raise RuntimeError()
    time.sleep(1)

@op(ins={"start": In(Nothing)})
def end():
    time.sleep(1)

@graph(out={"etl_1_test_graph_asset": GraphOut()})
def etl_1_test_graph_asset(r_etl_1_test_graph_asset):
    r_test_sleep_1 = test_sleep_1(r_etl_1_test_graph_asset)
    r_end = end(r_test_sleep_1)
    return {"etl_1_test_graph_asset": r_end}

r_etl_1_test_graph_asset = AssetsDefinition.from_graph(
    etl_1_test_graph_asset,
    keys_by_input_name={"r_etl_1_test_graph_asset": AssetKey("etl_1_test_graph_asset")},
    partition_mappings={"r_etl_1_test_graph_asset": TimeWindowPartitionMapping(start_offset=-1, end_offset=-1)},
    partitions_def=HourlyPartitionsDefinition("2023-01-22-18:00")
)

@repository
def test_repo():
    return [r_etl_1_test_graph_asset]
```
<img width="1004" alt="image" src="https://github.com/dagster-io/dagster/assets/42205607/3c6851b2-eece-4d0c-b5db-0ee892042ea8">

Please notice that I removed the code for the asset reconciliation sensor, because I noticed that it's marked as deprecated. So in this code snippet, it only defines a graph backed asset which has self past dependency. And here are my findings and questions:

- Even though from the UI, it says 'This asset depends on earlier partitions of itself.', when previous partition is un-materialized/ failed, I can still materialize the current partition. Why is that?
- If asset reconciliation sensor is deprecated, is there replacement for that? Are auto-materialise policy and freshness policy the replcacement?

**Reply by sryza** [reply]
> Even though from the UI, it says 'This asset depends on earlier partitions of itself.', when previous partition is un-materialized/ failed, I can still materialize the current partition. Why is that?

This issue is tracked here: https://github.com/dagster-io/dagster/issues/13896. FWIW, the partition won't be auto-materialized in this case.

> If asset reconciliation sensor is deprecated, is there replacement for that? Are auto-materialise policy and freshness policy the replcacement?

Exactly. Here are the docs for this: https://docs.dagster.io/concepts/assets/asset-auto-execution. If you want downstream partitions to be materialized as soon as upstream partitions are available, you can use eager auto-materialize policies, and you won't need freshness policies.

**Reply by renzhe-brian** [reply]
@sryza , thank you.  

In our environment, we use GraphDefinition to programmatically construct the dependencies for a graph. Therefore, I'm trying to convert the decorated function `etl_1_test_graph_asset`  to 
```python
def etl_1_test_graph_asset():
    return GraphDefinition(
        name="etl_1_test_graph_asset",
        node_defs=[test_sleep_1, end],
        input_mappings=[InputMapping("r_etl_1_test_graph_asset", "r_etl_1_test_graph_asset", "r_etl_1_test_graph_asset")],
        dependencies={"end": {"start": DependencyDefinition("test_sleep_1")}},
    )
```

However, this gives me error: 
`dagster._core.errors.DagsterInvalidDefinitionError: In GraphDefinition 'r_etl_1_test_graph_asset' input mapping references node 'r_etl_1_test_graph_asset' which it does not contain.`

Which kind of makes sense, because the asset `r_etl_1_test_graph_asset` is defined after this graph. How should I make it work the same as : 
```python
@graph(out={"etl_1_test_graph_asset": GraphOut()})
def etl_1_test_graph_asset(r_etl_1_test_graph_asset):
  ...
```

**Reply by sryza** [reply]
@renzhe-brian I'm having a little bit of trouble following. What exactly is your goal? Is it that you want to make a graph-backed asset with a self-partition-dependency?

Here's an example:

```python
from dagster import (
    graph_asset,
    op,
    In,
    Nothing,
    DailyPartitionsDefinition,
    AssetIn,
    TimeWindowPartitionMapping,
)


@op(ins={"in1": In(dagster_type=Nothing)})
def op1():
    ...


@graph_asset(
    partitions_def=DailyPartitionsDefinition(start_date="2023-05-10"),
    ins={
        "asset1": AssetIn(
            partition_mapping=TimeWindowPartitionMapping(start_offset=-1, end_offset=-1),
            dagster_type=Nothing,
        )
    },
)
def asset1(asset1):
    return op1(asset1)
```

**Reply by renzhe-brian** [reply]
Hi @sryza , you are right, that's my goal. The issue we have here is that in our environment we use `GraphDefinition` to programmatically construct the dependencies for a graph and then use it to make a graph-backed asset with `AssetsDefinition.from_graph(...)` . We can't use the decorator way of defining a graph-backed asset.
Here's my complete code snippet and it doesn't work:
```python
import time

from dagster import (
    In, Nothing, op, AssetsDefinition, AssetKey, TimeWindowPartitionMapping,
    HourlyPartitionsDefinition, repository, GraphDefinition, DependencyDefinition,
    InputMapping
)


@op(ins={"start": In(Nothing)})
def test_sleep_1():
    # if random.random() > 0.2:
    #     raise RuntimeError()
    time.sleep(1)

@op(ins={"start": In(Nothing)})
def end():
    time.sleep(1)

def etl_1_test_graph_asset():
    return GraphDefinition(
        name="etl_1_test_graph_asset",
        node_defs=[test_sleep_1, end],
        input_mappings=[InputMapping("r_etl_1_test_graph_asset", "r_etl_1_test_graph_asset", "r_etl_1_test_graph_asset")],
        dependencies={"end": {"start": DependencyDefinition("test_sleep_1")}},
    )


r_etl_1_test_graph_asset = AssetsDefinition.from_graph(
    etl_1_test_graph_asset(),
    keys_by_input_name={"r_etl_1_test_graph_asset": AssetKey("etl_1_test_graph_asset")},
    partition_mappings={"r_etl_1_test_graph_asset": TimeWindowPartitionMapping(start_offset=-1, end_offset=-1)},
    partitions_def=HourlyPartitionsDefinition("2023-01-22-18:00")
)

@repository
def test_repo():
    return [r_etl_1_test_graph_asset]

```

**Reply by sryza** [reply]
Here's an example with a `GraphDefinition`, using the same op as above:

```python
def make_graph_def():
    return GraphDefinition(
        name="graph1",
        input_mappings=[InputMapping("graph_in1", "op1", "in1")],
        output_mappings=[OutputMapping("graph_out1", "op1", "result")],
        node_defs=[op1],
    )


asset2 = AssetsDefinition.from_graph(
    make_graph_def(),
    partitions_def=DailyPartitionsDefinition(start_date="2023-05-10"),
    keys_by_input_name={"graph_in1": AssetKey("asset2")},
    partition_mappings={
        "graph_in1": TimeWindowPartitionMapping(start_offset=-1, end_offset=-1),
    },
)
```

**Reply by renzhe-brian** [reply]
I just tested and it works well. Thank you @sryza !

**Reply by renzhe-brian** [reply]
Hi @sryza , after testing this a bit further. I found some other issue.  Below is a code snippet that defines a graph-backed asset with a self-partition-dependency, and it works well. There're two ops in this graph: `op1` runs first then `end`.  However, `asset_key = context.selected_asset_keys` is None in `op1`. Is this a bug, or am I doing something wrong here?

```python
import time

from dagster import (
    In, Nothing, op, AssetsDefinition, AssetKey, TimeWindowPartitionMapping,
    HourlyPartitionsDefinition, repository, GraphDefinition, DependencyDefinition,
    InputMapping, OutputMapping, OpExecutionContext,DailyPartitionsDefinition,Out
)

@op(ins={"start": In(Nothing)}, out={"result": Out(Nothing)})
def op1(context: OpExecutionContext):
    asset_key = context.selected_asset_keys
    time.sleep(1)


@op(ins={"start": In(Nothing)}, out={"result": Out(Nothing)})
def end(context: OpExecutionContext):
    time.sleep(1)


def my_graph_asset():
    return GraphDefinition(
        name="my_graph_asset",
        node_defs=[op1, end],
        input_mappings=[InputMapping("my_graph_asset", "op1", "start")],
        output_mappings=[OutputMapping("result", "end", "result")],
        dependencies={"end": {"start": DependencyDefinition("op1")}},
    )


my_graph_asset = AssetsDefinition.from_graph(
    my_graph_asset(),
    keys_by_input_name={"my_graph_asset": AssetKey("my_graph_asset")},
    keys_by_output_name={"result": AssetKey("my_graph_asset")},
    partitions_def=DailyPartitionsDefinition(start_date="2023-06-10"),
    partition_mappings={"my_graph_asset": TimeWindowPartitionMapping(start_offset=-1,
                                                                           end_offset=-1)},
)

@repository
def test_repo():
    return [my_graph_asset]


```

**Comment by Jaapel** [comment]
Hi @sryza @OwenKephart .

I am working on a similar issue, but the ops in my asset graph do not return any output. I have a few different ops that together update a remote dataset, so I have a graph with a few different parallel ops. I wrote an example based on your suggestions:

```python
from dagster import (
    graph_asset,
    op,
    In,
    Nothing,
    DailyPartitionsDefinition,
    AssetIn,
    TimeWindowPartitionMapping,
)


@op(ins={"in1": In(dagster_type=Nothing)})
def op1() -> None:
    print("op1")


@op(ins={"in1": In(dagster_type=Nothing)})
def op2() -> None:
    print("op2")


@graph_asset(
    partitions_def=DailyPartitionsDefinition(start_date="2023-05-10"),
    ins={
        "asset1": AssetIn(
            partition_mapping=TimeWindowPartitionMapping(
                start_offset=-1, end_offset=-1
            ),
            dagster_type=Nothing,
        )
    },
)
def asset1(asset1) -> None:
    op1(asset1)
    op2(asset1)
```

When executing this code I get the following `CheckError`:
```
Invariant failed. Description: The set of output names keys specified in the keys_by_output_name argument must equal the set of asset keys outputted by asset1. 
keys_by_output_name keys: {'result'} 
expected keys: set()
```

Is this a bug or am I missing something?

**Reply by Jaapel** [reply]
I am also on dagster 1.3.9

**Reply by renzhe-brian** [reply]
@Jaapel , I think you can add an `end` op, which does nothing. And do it like this 
```python
@op(ins={"in1": In(dagster_type=Nothing), "in2": In(dagster_type=Nothing)}, out={"result": Out(Nothing)})
def end() -> None:
    print("do nothing")


@graph_asset(
    partitions_def=DailyPartitionsDefinition(start_date="2023-05-10"),
    ins={
        "asset1": AssetIn(
            partition_mapping=TimeWindowPartitionMapping(start_offset=-1, end_offset=-1),
            dagster_type=Nothing,
        )
    },

)
def asset1(asset1):
    return end(in1=op1(asset1), in2=op2(asset1))
```

I'm not sure if this is the proper fix though.

**Reply by OwenKephart** [reply]
The above fix is what I'd recommend as well! Even if there's no physical output from the op, your setup is that asset1 will only be updated once both op1 and op2 have finished, so if we were to tie the materialization of asset1 to any particular event, this is the best way of doing it (essentially this `end` op lets you get an output that represents "both op1 and op2 finished")

**Reply by Jaapel** [reply]
This is great guys, thanks for the quick reply
